sw/device/lib/crypto/drivers/hmac.c

```diff
@@ -256,6 +225,120 @@ static void msg_fifo_write(const uint8_t *message, size_t message_len) {
+  // will unlock us from the state fifo_st_q==FifoLoadFromFifo in the block
+  // prim_sha2.sv.
+  uint32_t cmd_reg = abs_mmio_read32(kHmacBaseAddr + HMAC_CMD_REG_OFFSET);
+  cmd_reg = bitfield_bit32_write(cmd_reg, HMAC_CMD_HASH_CONTINUE_BIT, true);
+  abs_mmio_write32(kHmacBaseAddr + HMAC_CMD_REG_OFFSET, cmd_reg);
+
+  // Get the current message length to know how much words we need to write to
+  // fall on the block boundary and trigger digest_on_blk to be able to move
+  // into done_state_d==DoneAwaitCmd in hmac.sv this will then lead to trigger
+  // hash_done_event and be back in a stable state on all the FSMs.
+  uint64_t msg_len = ((uint64_t)ctx->upper << 32) | ctx->lower;
+  uint32_t digest_size =
+      bitfield_field32_read(cfg_reg, HMAC_CFG_DIGEST_SIZE_FIELD);
+
+  // Compute next block boundary
+  uint32_t msg_length_to_wr;
+  // SHA2-256 mode
+  if (digest_size == 1) {
+    msg_length_to_wr = kHmacSha256BlockBits - msg_len % kHmacSha256BlockBits;
+  } else {
+    msg_length_to_wr = kHmacSha512BlockBits - msg_len % kHmacSha512BlockBits;
+  }
+
+  // Write a dummy message into the message FIFO to trigger shaf_rvalid_o
+  // from prim_sha2_pad.sv
+  for (int i = 0; i < msg_length_to_wr / 32; i++) {
+    abs_mmio_write32(kHmacBaseAddr + HMAC_MSG_FIFO_REG_OFFSET, 0xFFDEADFF);
+  }
+
+  // Finally trigger hash_process
+  cmd_reg = abs_mmio_read32(kHmacBaseAddr + HMAC_CMD_REG_OFFSET);
+  cmd_reg = bitfield_bit32_write(cmd_reg, HMAC_CMD_HASH_PROCESS_BIT, true);
+  abs_mmio_write32(kHmacBaseAddr + HMAC_CMD_REG_OFFSET, cmd_reg);
+
+  // Restore CFG.hmac_en as it was before
+  cfg_reg = bitfield_bit32_write(cfg_reg, HMAC_CFG_HMAC_EN_BIT, hmac_en);
+  abs_mmio_write32(kHmacBaseAddr + HMAC_CFG_REG_OFFSET, cfg_reg);
+}
+
+/**
+ * Wait until HMAC becomes idle.
+ *
+ * It returns error if HMAC HWIP becomes idle without firing `hmac_done`
+ * interrupt.
+ *
+ * TODO(#23191): It might be beneficial to have a timeout value for the polling.
+ *
+ * @param[out] ctx Context to which values are written.
+ */
+static status_t hmac_idle_wait(hmac_ctx_t *ctx) {
+  status_t status;
+  // Verify that HMAC HWIP is idle.
+  // Initialize `status_reg = 0` so that the loop starts with the assumption
+  // that HMAC HWIP is not idle.
+  // TODO: should be removed when issue #24767 will be solved in the HW
+  //  | At max 4 clock cycles are required to perform a read access to the
+  //  | register. As it should take less than 64 clock cycles in SHA2-256 and
+  //  | 80 clock cycles in SHA2-384/512, let's take some margin and consider
+  //  | that 100 loops are a way enough to see IDLE status. Otherwise we can
+  //  | start to attempt to recover the HW.
+  uint32_t status_reg = 0;
+  uint32_t attempt_cnt = 0;
+  while (bitfield_bit32_read(status_reg, HMAC_STATUS_HMAC_IDLE_BIT) == 0) {
+    status_reg = abs_mmio_read32(kHmacBaseAddr + HMAC_STATUS_REG_OFFSET);
+    attempt_cnt++;
+    if (attempt_cnt == 100) {
+      recover_hw_after_stop(ctx);
```

Comment:
I think you could simply return `OTCRYPTO_RECOV_ERR` here and exit, and then look for that case in the function that calls this one and run the `recover_hw_after_stop()` operation there. Then you wouldn't need to pass `ctx` to this function at all.

Reply:
If I do so, the function will stop here and won't execute this remaining piece of code?
```
  // Verify that HMAC HWIP raises `hmac_done` bit.
  uint32_t intr_reg =
      abs_mmio_read32(kHmacBaseAddr + HMAC_INTR_STATE_REG_OFFSET);
  if (bitfield_bit32_read(intr_reg, HMAC_INTR_STATE_HMAC_DONE_BIT) == 0) {
    status = OTCRYPTO_FATAL_ERR;
  }

  // Clear the interrupt by writing 1, because `INTR_STATE` is rw1c type.
  abs_mmio_write32(kHmacBaseAddr + HMAC_INTR_STATE_REG_OFFSET, intr_reg);
  if (status.value != OTCRYPTO_RECOV_ERR.value) {
    status = OTCRYPTO_OK;
  }

  return status;
```
I think we should execute those read and write operations after the HW has been recovered or not.